### ISSUE TITLE
kvserver: replicated into unavailability

### ISSUE THREAD

**Issue by tbg** [cockroachdb/cockroach#64649]
**Describe the problem**

v20.2.8

Unavailable ranges in a failover test observed by @fabiog1901, conducted on behalf of a customer.

Three localities with three nodes in each

- us-central1-c
- us-central1-f
- us-east4

Initially replication factor 5 with constraints: 2 replicas in each central locality, one in us-east4.

Want to test what happens when a central locality goes down: take that DC down, wait until the nodes are recognized as dead (5 minutes), then change the replication factor to three with constraints

```
num_replicas = 3,
	constraints = '{+zone=us-central1-f: 2, +region=us-east4: 1}',
	lease_preferences = '[[+region=us-central1]]';
```

Soon, unavailable ranges are observed. In the code block below, n1-n3 are in the down `us-central1-c` region. We can see that r844 initially had five voters on nodes n1 n2 n4 n5 n9. Node that n1 and n2 are down and that they have been non-live for minutes by the time this replication change occurs (which is presumably right after the replication factor and constraints are changed five minutes into the test). So what we're seeing is that a downreplication is initiated that loses quorum: from (n1(down) n2(down) n4 n5 n9) we go to (n1(down) n2(down) n4(demoting) n5 n9), which requires a quorum over (n1 n2 n5 n9) to make progress. But this configuration has a quorum size of three, and only two nodes within it are live.

```
2021-05-03 15:40:13	s68	Remove	r844		

    Updated Range Descriptor: r844:/Table/66/4/{2/18750/"CHIPS"/"C"/"I-\xfb\r\xae\x8cJ\xbe\x90\x003\v&\"\xe4X"/"EVENT-10"-4/18750/"CHIPS"/"C"/"\x86\x82\xe9\x1a\xc5\xfeM\xa6\xaeD\xae3\xa1\xc0\xa7S"/"EVENT-6"} [(n5,s40):1VOTER_DEMOTING, (n1,s1):2, (n2,s9):3, (n4,s28):4, (n9,s68):5, next=6, gen=487, sticky=1620055700.436036148,0]
    Removed Replica: (n5,s40):1VOTER_DEMOTING
    Reason: admin request
```

The "admin request" hints at this being a side effect of a store rebalance, which should never happen. This reminds me a lot of https://github.com/cockroachdb/cockroach/pull/56735. Probably this code

https://github.com/cockroachdb/cockroach/blob/b00f9239d3934a60eac3aaf443b1c94c082fe9a0/pkg/kv/kvserver/store_rebalancer.go#L542-L558

suddenly needs fewer voters than there are in the descriptor, and so the rebalance turns into an unguided downreplication. I remember seeing this in the past, and thought we had fixed it, though the link above is to the master branch and doesn't show signs of a fix (though a lower-level check would prevent it I believe).

**To Reproduce**

Follow the steps above after cross-checking with [this thread](https://cockroachlabs.slack.com/archives/CHVV403F0/p1620058136264600)

**Expected behavior**

We should not attempt or even carry out replication changes that jeopardize range availability.

**Additional data / screenshots**

**Environment:**
 - CockroachDB version 20.2.8

**Additional context**

On `master`, we have a low-level check against dangerous replication changes:

https://github.com/cockroachdb/cockroach/blob/1fc0e3d028cecead0e6671d20f560f78e11a0f94/pkg/kv/kvserver/replica_command.go#L2129-L2144

We weren't able to backport this to 20.2 at the time, but here it looks like we've found a real-world repro of https://github.com/cockroachdb/cockroach/pull/56735 and need to develop a fix regardless.

**Comment by tbg** [comment]
https://github.com/cockroachdb/cockroach/issues/54444#issuecomment-707706553 is the smoking gun - pretty sure this will have been the same thing. I think there is a "clear" fix, which is making sure that the store rebalancer does not accidentally change the number of voters (or non-voters for that matter).

**Comment by fabiog1901** [comment]
Having tested on the patch version, as soon as I proceeded with destroying dc `us-central1-c`, I got 1 Unavailable Range. I can confirm that, prior to destroying the DC, there was no notification of any UnderReplicated Range.

After reviewing the Range Report for that range, it seems that range was only replicated 4 times as for some unknown reason, it wasn't able to upreplicate to 5 replicas:


Timestamp | Store | Event Type | Range | Other Range | Info
-- | -- | -- | -- | -- | --
2021-05-05 14:39:42 | s67 | Remove | r28010 |   | Updated Range Descriptor: r28010:/Table/66/1/"o\xb{6\xb5\xf0\xdd!M\x04\xac\xc0PV\xb8\xfc[\|"/"EVENT-7"-c\f2W\xa3O \x9dE\xa1\xfc]\xac\x9b\x83"/"EVENT-7"/0} [(n2,s11):1, (n4,s25):2, (n9,s67):3, (n3,s22):7, next=9, gen=509, sticky=1620225026.436929780,0]Removed Replica: (n6,s46):8LEARNERReason: abandoned learner replica
2021-05-05 14:39:33 | s67 | Add | r28010 |   | Updated Range Descriptor: r28010:/Table/66/1/"o\xb{6\xb5\xf0\xdd!M\x04\xac\xc0PV\xb8\xfc[\|"/"EVENT-7"-c\f2W\xa3O \x9dE\xa1\xfc]\xac\x9b\x83"/"EVENT-7"/0} [(n2,s11):1, (n4,s25):2, (n9,s67):3, (n3,s22):7, (n6,s46):8LEARNER, next=9, gen=508, sticky=1620225026.436929780,0]Added Replica: (n6,s46):8LEARNERReason: range under-replicatedDetails: {"Target":"s46, necessary:true, diversity:0.50, converges:0, balance:-1, rangeCount:1392"}

So as soon as n1, n2, n3 were destroyed, there was no quorum


Timestamp | Message
-- | --
2021-05-05 15:57:49 | kv/kvserver/store.go:2711 [n9,status] error simulating allocator on replica [n9,s67,r28010/3:/Table/66/1/"o\xb{6\xb…-c\f2…}]: range requires a replication change, but live replicas [(n4,s25):2 (n9,s67):3] don't constitute a quorum for [(n2,s11):1 (n4,s25):2 (n9,s67):3 (n3,s22):7]:

**Comment by fabiog1901** [comment]
FWIW, I run the same experiment 3 times, and in 2 occasions I was able to replicate the above. Interestingly, it's always the same range `/Table/66/1/`. If helpful, the dataset is added to the cluster by means of a Restore job. Happy to assist replicating the issue.

**Comment by aayushshah15** [comment]
> before the replicate queue got around to this range again

One thing that's struck me as odd in the recent past is that, [_on any kind of failure when performing a replication change_](https://github.com/cockroachdb/cockroach/blob/7495434b547ce1e832d27ebedb3e742e7a254fea/pkg/kv/kvserver/replicate_queue.go#L644-L655), we return `requeue: false`. Maybe we can sniff the type of error and conditionally requeue (up to some reasonable number of retries) immediately when we know it's a retryable failure?

**Comment by lunevalex** [comment]
@aayushshah15 thats a good point, looking at the report above its a pretty huge cluster i.e the range number is 28010 on Store 67, it would take the replicate_queue quite awhile to swing around and try to upreplicate again.

**Comment by tbg** [comment]
Fixes landed back in May, so not sure there's anything left here?

master: https://github.com/cockroachdb/cockroach/pull/64650
20.2: https://github.com/cockroachdb/cockroach/pull/64652
21.1: https://github.com/cockroachdb/cockroach/pull/65501

**Comment by BramGruneir** [comment]
@tbg, is there a need to add a test?

**Comment by tbg** [comment]
Certainly! All of the StoreRebalancer needs tons of tests. Unfortunately this code is very good at resisting meaningful testing (we're sad about that) and it's highly unlikely that this issue will lead to a test being written. https://github.com/cockroachdb/cockroach/issues/70552 will hopefully help here and there is desire to work on this though we'll see how it fares in 22.1 planning.
